File: frontend/apps/chinese/src/features/srs/lib/sessionDeck.test.ts
import { describe, expect, it } from 'vitest'
import { clampDurationMs, countRatings, formatSessionDuration, mergeIncoming, shouldLoadMore } from './sessionDeck'
import type { SrsCardState, SrsQueueCard } from '../types'

const card = (cardId: string, state: SrsCardState = 'review'): SrsQueueCard => ({
  cardId,
  state,
  queue: state === 'new' ? 'new' : state === 'review' ? 'review' : 'learning',
  dueAt: '2026-09-18T00:00:00Z',
  word: { id: 'w-' + cardId, simplified: '爱', pinyin: 'ai4', meaningViStatus: 'machine' },
  intervals: { again: 'PT1M', hard: 'PT5M30S', good: 'PT10M', easy: 'P8D' },
})

describe('mergeIncoming', () => {
  it('bỏ thẻ đã có trong bộ (chưa chấm)', () => {
    const deck = [card('a'), card('b')]
    const out = mergeIncoming(deck, new Set(), new Set(), [card('b'), card('c')])
    expect(out.map((c) => c.cardId)).toEqual(['c'])
  })

  it('thẻ đã chấm chỉ quay lại khi server trả learning/relearning', () => {
    // Cả a lẫn b đã chấm ⇒ phần chưa chấm rỗng.
    const rated = new Set(['a', 'b'])
    const out = mergeIncoming([], rated, new Set(), [card('a', 'learning'), card('b', 'review'), card('c', 'relearning')])
    expect(out.map((c) => c.cardId)).toEqual(['a', 'c'])
  })

  it('bỏ thẻ đang chờ gửi trong outbox', () => {
    const out = mergeIncoming([], new Set(['a']), new Set(['a']), [card('a', 'learning'), card('b')])
    expect(out.map((c) => c.cardId)).toEqual(['b'])
  })

  it('bản quay lại của thẻ đã chấm đang nằm trong phần chưa chấm ⇒ không nối thêm lần nữa (review F7)', () => {
    // Bộ bài [a, b, a(learning)], index = 1 ⇒ phần chưa chấm = [b, a(learning)], rated = {a}.
    const deck = [card('a', 'new'), card('b'), card('a', 'learning')]
    const unrated = deck.slice(1)
    const out = mergeIncoming(unrated, new Set(['a']), new Set(), [card('a', 'learning')])
    expect(out).toEqual([])
  })

  it('skipNew ⇒ bỏ thẻ new (server đếm từ mới chưa kịp khi còn lượt chấm đang bay) — tích hợp F7', () => {
    const incoming = [card('n1', 'new'), card('r1', 'review'), card('l1', 'learning')]
    expect(mergeIncoming([], new Set(['p']), new Set(['p']), incoming, true).map((c) => c.cardId)).toEqual(['r1', 'l1'])
    // Không bật cờ (tải lần đầu / không có lượt đang bay) ⇒ không lọc.
    expect(mergeIncoming([], new Set(), new Set(['p']), incoming).map((c) => c.cardId)).toEqual(['n1', 'r1', 'l1'])
  })

  it('không thêm trùng trong cùng một lô tải', () => {
    const out = mergeIncoming([], new Set(), new Set(), [card('x'), card('x')])
    expect(out).toHaveLength(1)
  })
})

describe('shouldLoadMore', () => {
  it('còn ≤ 5 thẻ, không đang tải, chưa hết ⇒ tải', () => {
    expect(shouldLoadMore(5, false, false)).toBe(true)
    expect(shouldLoadMore(0, false, false)).toBe(true)
    expect(shouldLoadMore(6, false, false)).toBe(false)
    expect(shouldLoadMore(2, true, false)).toBe(false)
    expect(shouldLoadMore(2, false, true)).toBe(false)
  })
})

describe('countRatings / formatSessionDuration / clampDurationMs', () => {
  it('đếm theo 4 mức', () => {
    expect(countRatings(['good', 'again', 'good', 'easy'])).toEqual({ again: 1, hard: 0, good: 2, easy: 1 })
  })

  it('định dạng thời gian phiên', () => {
    expect(formatSessionDuration(45_000)).toBe('45 giây')
    expect(formatSessionDuration(200_000)).toBe('3 phút 20 giây')
    expect(formatSessionDuration(120_000)).toBe('2 phút')
    expect(formatSessionDuration(3_720_000)).toBe('1 giờ 2 phút')
    expect(formatSessionDuration(-5)).toBe('0 giây')
  })

  it('kẹp durationMs 0..600000', () => {
    expect(clampDurationMs(-10)).toBe(0)
    expect(clampDurationMs(4200.6)).toBe(4201)
    expect(clampDurationMs(999_999)).toBe(600_000)
    expect(clampDurationMs(Number.NaN)).toBe(0)
  })
})

File: mobile/apps/chinese/test/features/srs/session_deck_test.dart
import 'package:af_chinese/features/srs/data/models.dart';
import 'package:af_chinese/features/srs/domain/session_deck.dart';
import 'package:flutter_test/flutter_test.dart';

// Chép đủ ca của `sessionDeck.test.ts` web.
SrsQueueCard card(String cardId, [SrsCardState state = SrsCardState.review]) => SrsQueueCard(
  cardId: cardId,
  state: state,
  queue: state == SrsCardState.fresh
      ? 'new'
      : state == SrsCardState.review
      ? 'review'
      : 'learning',
  dueAt: DateTime.utc(2026, 9, 18),
  word: SrsQueueWord(id: 'w-$cardId', simplified: '爱', pinyin: 'ai4', meaningViStatus: 'machine'),
  intervals: const {SrsRating.again: 'PT1M', SrsRating.hard: 'PT5M30S', SrsRating.good: 'PT10M', SrsRating.easy: 'P8D'},
);

List<String> ids(List<SrsQueueCard> cards) => cards.map((c) => c.cardId).toList();

void main() {
  group('mergeIncoming', () {
    test('bỏ thẻ đã có trong bộ (chưa chấm)', () {
      final deck = [card('a'), card('b')];
      expect(ids(mergeIncoming(deck, {}, {}, [card('b'), card('c')])), ['c']);
    });

    test('thẻ đã chấm chỉ quay lại khi server trả learning/relearning', () {
      final out = mergeIncoming(
        [],
        {'a', 'b'},
        {},
        [card('a', SrsCardState.learning), card('b', SrsCardState.review), card('c', SrsCardState.relearning)],
      );
      expect(ids(out), ['a', 'c']);
    });

    test('bỏ thẻ đang chờ gửi trong outbox', () {
      final out = mergeIncoming([], {'a'}, {'a'}, [card('a', SrsCardState.learning), card('b')]);
      expect(ids(out), ['b']);
    });

    test('bản quay lại của thẻ đã chấm đang nằm trong phần chưa chấm ⇒ không nối thêm lần nữa (review F7)', () {
      final deck = [card('a', SrsCardState.fresh), card('b'), card('a', SrsCardState.learning)];
      final unrated = deck.sublist(1);
      expect(mergeIncoming(unrated, {'a'}, {}, [card('a', SrsCardState.learning)]), isEmpty);
    });

    test('skipNew ⇒ bỏ thẻ new (server đếm từ mới chưa kịp khi còn lượt chấm đang bay) — tích hợp F7', () {
      final incoming = [card('n1', SrsCardState.fresh), card('r1'), card('l1', SrsCardState.learning)];
      expect(ids(mergeIncoming([], {'p'}, {'p'}, incoming, skipNew: true)), ['r1', 'l1']);
      // Không bật cờ (tải lần đầu / không có lượt đang bay) ⇒ không lọc.
      expect(ids(mergeIncoming([], {}, {'p'}, incoming)), ['n1', 'r1', 'l1']);
    });

    test('không thêm trùng trong cùng một lô tải', () {
      expect(mergeIncoming([], {}, {}, [card('x'), card('x')]), hasLength(1));
    });
  });

  group('shouldLoadMore', () {
    test('còn ≤ 5 thẻ, không đang tải, chưa hết ⇒ tải', () {
      expect(shouldLoadMore(5, loading: false, exhausted: false), isTrue);
      expect(shouldLoadMore(0, loading: false, exhausted: false), isTrue);
      expect(shouldLoadMore(6, loading: false, exhausted: false), isFalse);
      expect(shouldLoadMore(2, loading: true, exhausted: false), isFalse);
      expect(shouldLoadMore(2, loading: false, exhausted: true), isFalse);
    });
  });

  group('countRatings / formatSessionDuration / clampDurationMs', () {
    test('đếm theo 4 mức', () {
      expect(countRatings([SrsRating.good, SrsRating.again, SrsRating.good, SrsRating.easy]), {
        SrsRating.again: 1,
        SrsRating.hard: 0,
        SrsRating.good: 2,
        SrsRating.easy: 1,
      });
    });

    test('định dạng thời gian phiên', () {
      expect(formatSessionDuration(45000), '45 giây');
      expect(formatSessionDuration(200000), '3 phút 20 giây');
      expect(formatSessionDuration(120000), '2 phút');
      expect(formatSessionDuration(3720000), '1 giờ 2 phút');
      expect(formatSessionDuration(-5), '0 giây');
    });

    test('kẹp durationMs 0..600000', () {
      expect(clampDurationMs(-10), 0);
      expect(clampDurationMs(4200.6), 4201);
      expect(clampDurationMs(999999), 600000);
      expect(clampDurationMs(double.nan), 0);
    });
  });
}
